django/db/models/indexes.py

```diff
@@ -0,0 +1,78 @@
+
+__all__ = ['Index']
+
+
+class BaseIndex(object):
+
+    # The type of index - BTree, Hash, RTree, etc.
+    # Used in deconstruction to differentiate indexes of different types with
+    # the same set of fields.
+    index_type = None
+
+    def __init__(self, *fields, **kwargs):
+        if not fields:
+            raise ValueError("Minimum one field required to define an index.")
+        self.fields = fields
+        if 'model' in kwargs:
+            self.model = kwargs['model']
+        if 'name' in kwargs:
+            self.name = kwargs['name']
```

Comment:
I'm wondering if it makes sense to have `self._name` default to `None` and a property `name` that either returns `self._name` or automatically generates one and stores it on `self._name` which is then returned.